Allow multiple FlexibleAreas in a FlexibleStopPlace

hannesj opened the pull request:
### Summary

Currently only the first FlexibleArea in a NeTEX FlexibleStopPlace is parsed. Modify the code to create a GrupStop if multiple areas exist.

jtorin commented on src/main/java/org/opentripplanner/netex/mapping/FlexStopsMapper.java:
```diff
@@ -58,43 +58,57 @@ class FlexStopsMapper {
-          .stream()
-          .filter(k -> k.getKey().equals(FLEXIBLE_STOP_AREA_TYPE_KEY))
-          .findFirst();
+      String flexibleAreaType = null;
```
The following code in the loop doesn't use `area` , only `flexibleStopPlace` directly. Effectively the method always returns on the first iteration, leading to not all areas being type checked above.

Suggestion is to move the type check to a separate method or at least move out the following code from the loop.

hannesj replied:
This is a question. I was under the impression that the key should be in the FlexibleArea, not the FlexibleStopPlace, as that makes it impossible to combine different types of areas inside a stop place, which is a goal of ours.

testower replied:
I think the type key/value should be possible to specify on both FlexibleStopPlace and FlexibleArea, but if there it is on FlexibleArea, then that should have priority. Then we have backwards compatibility with existing implementations.

t2gran commented:
I asked @hannesj if we(me or @vpaturet) should take over this one. I agree with the comments from @jtorin.

hannesj commented:
I addressed most of the comments. There is an open question about the location of the type key (see [comment](https://github.com/opentripplanner/OpenTripPlanner/pull/4922#discussion_r1126921612)).

leonardehrenfried commented:
Could you please fix the tests?

t2gran commented on src/main/java/org/opentripplanner/netex/mapping/FlexStopsMapper.java:
```diff
@@ -59,85 +57,82 @@ class FlexStopsMapper {
+            area
+          )
+        );
+        LOG.warn(
+          "FlexibleStopPlace {} contains an unsupported area {}. Hail and ride areas are not currently supported.",
+          flexibleStopPlace.getId(),
+          area
+        );
```
When an issue is added to the issue store it is logged with `debug` level. We could possibly change this, but I do not think we should log here. If we want to log different issues with different log level we should add extra methods to the `DataImportIssueStore` e.g. `addAndLogWarn`, `addAndLogInfo` and `addAndLogDebug`. Another but too me better solution is to inside the `DataImportIssueStore` to create a logger per issue type (o.o.issue.<issue-type>) and log as info. Then each deployment can control the logging as they like.

I can do this in another PR - just want to discuss it in a dev meeting first. In this PR i suggest just removing the log statment.

testower replied:
Should we remove the other 2 LOG.warn statements as well?

hannesj commented:
> Could you please fix the tests?

No, as the tests should fail with the current behaviour and data. We need to clarify the https://github.com/opentripplanner/OpenTripPlanner/pull/4922#discussion_r1126921612 above

vpaturet reviewed (changes requested):
As suggested by @testower, the key/value pairs should be looked up both at the FlexibleStopPlace level and FlexibleArea level, with precedence at the FlexibleArea level.

testower commented on src/main/java/org/opentripplanner/netex/mapping/FlexStopsMapper.java:
```diff
@@ -117,25 +131,22 @@ AreaStop mapFlexArea(FlexibleStopPlace flexibleStopPlace, FlexibleArea area) {
     Geometry geometry = mapGeometry(area);
     if (geometry == null) {
-      return null;
+      return List.of();
```
I can't tell what the significance of this change is. @hannesj would you be able to shine some light?

hannesj replied:
This makes the build process not crash, if there is an stop area with an invalid geometry. Otherwise iteration of null would cause an NPE.

testower replied:
Yeah, so trying to find the stops within an area with invalid geometry will result in an empty list of stops, whereas before it would cause an NPE and stop the whole build? Do we even allow FlexibleArea's with invalid geometry to get this far or should it be stopped earlier in the process?

hannesj replied:
This is the place where the `OpenGisMapper` is called, and the first place where geometry issues, as long as the geometry can be deserialized, are uncovered.

leonardehrenfried commented:
@testower Can you please click "re-request review" when you think that it's time for it?

testower commented:
> @testower Can you please click "re-request review" when you think that it's time for it?

Yep will do, there's a couple of outstanding issues that I'm trying to resolve first.

testower commented:
I have verified this locally with a semi-fictitious dataset (I have excluded the regular stops from this zip, otherwise GitHub won't let me upload it):

[test.zip](https://github.com/opentripplanner/OpenTripPlanner/files/11824388/test.zip)

The dataset contains a flexible service that uses a single flexible stop place, with two areas. One of the areas is of the type UnrestrictedPublicTransportArea. Here is screenshot from the debug client, show an itinerary from a regular stop within one area to a street edge in the other area:

![Screenshot 2023-06-21 at 21 56 30](https://github.com/opentripplanner/OpenTripPlanner/assets/231492/84166d90-db19-4bc1-8a6c-37b80997ca2f)

testower commented:
Should be good to go now @leonardehrenfried

leonardehrenfried commented:
Ok, I see that you've already have reviewers (of which I'm not one, because I don't use Netex :)).

leonardehrenfried reviewed (approved):
@testower Do you have merge permission?
